slack-lint-checks/src/main/java/slack/lint/CircuitScreenDataClassDetector.kt

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2025 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package slack.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.getUMethod
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtPrimaryConstructor
+import org.jetbrains.uast.UClass
+import org.jetbrains.uast.UElement
+import slack.lint.util.implements
+import slack.lint.util.sourceImplementation
+
+class CircuitScreenDataClassDetector : Detector(), SourceCodeScanner {
+  override fun getApplicableUastTypes(): List<Class<out UElement>> = listOf(UClass::class.java)
+
+  override fun createUastHandler(context: JavaContext): UElementHandler {
+    return object : UElementHandler() {
+
+      override fun visitClass(node: UClass) {
+        val sourceNode = node.sourcePsi as? KtClassOrObject ?: return
+
+        val isInterface = node.isInterface
+        // Open classes cannot be "data" classes
+        val isOpen = sourceNode.hasModifier(KtTokens.OPEN_KEYWORD)
+        // Screens must be parcelable and inner classes cannot be parcelable
+        val isInner = sourceNode.hasModifier(KtTokens.INNER_KEYWORD)
+        // Cannot have abstract data class / object
+        val isAbstract = sourceNode.hasModifier(KtTokens.ABSTRACT_KEYWORD)
+        // Cannot have sealed data class / object
+        val isSealed = sourceNode.hasModifier(KtTokens.SEALED_KEYWORD)
+        // Cannot have companion data object
+        val isCompanionObject = sourceNode.hasModifier(KtTokens.COMPANION_KEYWORD)
+
+        val isApplicableClass =
+          !isOpen && !isCompanionObject && !isInterface && !isInner && !isAbstract && !isSealed
+
+        if (isApplicableClass && node.implements(QUALIFIED_CIRCUIT_SCREEN)) {
+          val isDataClass = sourceNode.isData()
```

Comment:
nit: Looks like the `isDataClass` check could be an early return or be one of the `isApplicableClass`  values